src/python/pants/pantsd/service/scheduler_service.py

```diff
@@ -65,23 +74,48 @@ def setup(self, lifecycle_lock, fork_lock):
       )
     self._logger.info('watching invalidating files: {}'.format(self._invalidating_files))
 
+    if self._pidfile:
+      self._fs_event_service.register_pidfile_handler(self._pidfile, self._enqueue_fs_event)
+
   def _enqueue_fs_event(self, event):
     """Watchman filesystem event handler for BUILD/requirements.txt updates. Called via a thread."""
     self._logger.info('enqueuing {} changes for subscription {}'
                       .format(len(event['files']), event['subscription']))
     self._event_queue.put(event)
 
-  def _maybe_invalidate_scheduler(self, files):
+  def _maybe_invalidate_scheduler_batch(self, files):
     invalidating_files = self._invalidating_files
     if any(f in invalidating_files for f in files):
       self._logger.fatal('saw file events covered by invalidation globs, terminating the daemon.')
       self.terminate()
 
+  def _maybe_invalidate_scheduler_pidfile(self):
+    new_pid = self._check_pid_changed()
+    if new_pid is not False:
+      self._logger.fatal('{} says pantsd PID is {} but my PID is: {}: terminating'.format(
+        self._pidfile,
+        new_pid,
+        os.getpid(),
+      ))
+      self.terminate()
+
+  def _check_pid_changed(self):
+    """Reads pidfile and returns False if its PID is ours, else a printable (maybe falsey) value."""
+    try:
+      with open(os.path.join(self._build_root, self._pidfile), "r") as f:
+        pid_from_file = f.read()
+    except IOError:
+      return "[no file could be read]"
```

Comment:
if we can't read the pidfile after explicitly acquiring the lifecycle lock, its likely because it's been removed (by e.g. a rogue `git clean`) - and I think that should probably also result in a teardown here as well (e.g. `return False`)? what do you think?

Reply:
The `return False` case is the case where we _shouldn't_ tear down. The current behaviour here is to teardown if that file can't be read.